solr/core/src/test/org/apache/solr/cloud/overseer/ZkStateReaderTest.java

```diff
@@ -668,6 +673,81 @@ public void testWatchRaceCondition() throws Exception {
     }
   }
 
+  /**
+   * Simulates race condition that might arise when state updates triggered by watch notification
+   * contend with removal of collection watches.
+   *
+   * <p>Such race condition should no longer exist with the new code that uses a single map for both
+   * "collection watches" and "latest state of watched collection"
+   */
+  public void testStateWatcherRaceCondition() throws Exception {
+    ZkStateWriter writer = fixture.writer;
+    final ZkStateReader reader = fixture.reader;
+    fixture.zkClient.makePath(ZkStateReader.COLLECTIONS_ZKNODE + "/c1", true);
+    int extraWatchers = 10;
+    int iterations = 10;
+    for (int i = 0; i < extraWatchers; i++) {
+      // add and remove a bunch of watchers
+      DocCollectionWatcher w = (coll) -> false;
```

Comment:
I assume generally we want to return true here. Wondering what happens if we return true here?

Reply:
I think it depends. Some watches are one-off (like in `waitForState`), but there's at least one watcher that's persistent for the life of the core: `ZkController.UnloadCoreOnDeletedWatcher`. In this test, because the point is that we add-and-remove quickly, returning `true` wouldn't change anything ... it would just be another way of removing the watcher.

Reply:
We just want to make sure with prs collection and restart, we see this behavior?